galsim/phase_screens.py

```diff
@@ -0,0 +1,623 @@
+                   multiple of self.time_step.
+        """
+        if dt < 0:
+            raise ValueError("Cannot advance phase screen backwards in time.")
+        _nstep = int(np.round(dt/self.time_step))
+        if _nstep == 0:
+            _nstep = 1
+        for i in xrange(_nstep):
+            self.advance()
+        return _nstep*self.time_step  # return the time *actually* advanced
+
+    # Note -- use **kwargs here so that AtmosphericScreen.stepK and OpticalScreen.stepK
+    # can use the same signature, even though they depend on different parameters.
+    def stepK(self, **kwargs):
+        """Return an appropriate stepk for this atmospheric layer.
+
+        @param lam         Wavelength in nanometers.
+        @param scale_unit  Sky coordinate units of output profile. [Default: galsim.arcsec]
+        @param gsparams    An optional GSParams argument.  See the docstring for GSParams for
+                           details. [default: None]
+        @returns  Good pupil scale size in meters.
+        """
+        lam = kwargs['lam']
+        gsparams = kwargs.pop('gsparams', None)
+        obj = galsim.Kolmogorov(lam=lam, r0_500=self.r0_500, gsparams=gsparams)
+        return obj.stepK()
+
+    def wavefront(self, aper, theta_x=0.0*galsim.degrees, theta_y=0.0*galsim.degrees,
+                  compact=True):
+        """ Compute wavefront due to phase screen.
+
+        Wavefront here indicates the distance by which the physical wavefront lags or leads the
+        ideal plane wave (pre-optics) or spherical wave (post-optics).
+
+        @param aper     `galsim.Aperture` over which to compute wavefront.
+        @param theta_x  x-component of field angle corresponding to center of output array.
+        @param theta_y  y-component of field angle corresponding to center of output array.
+        @param compact  If true, then only return wavefront for illuminated pixels.
+        @returns        Wavefront lag or lead in nanometers over aperture.
+        """
+        if compact:
+            u, v = aper.u[aper.illuminated], aper.v[aper.illuminated]
+        else:
+            u, v = aper.u, aper.v
+        return self.tab2d(u + self.origin[0] + 1000*self.altitude*theta_x.tan(),
+                          v + self.origin[1] + 1000*self.altitude*theta_y.tan())
+
+    def reset(self):
+        """Reset phase screen back to time=0."""
+        self.rng = self.orig_rng.duplicate()
+        self.origin = np.array([0.0, 0.0])
+
+        # Only need to reset/create tab2d if not frozen or doesn't already exist
+        if self.alpha != 1.0 or not hasattr(self, 'tab2d'):
+            self.screen = self._random_screen()
+            self._xs = np.linspace(-0.5*self.screen_size, 0.5*self.screen_size, self.npix,
+                                   endpoint=False)
+            self._ys = self._xs
+            self.tab2d = galsim.LookupTable2D(self._xs, self._ys, self.screen, edge_mode='wrap')
+
+
+def _listify(arg):
```

Comment:
Any reason not to put this and the next utility routine in utilities.py so they can be more widely used?